### ISSUE TITLE
Pipeline default executables and environment substitution

### ISSUE THREAD

**Issue by toolmantim** [buildkite/agent#205]
Say you had the following `.buildkite/pipeline.template.yml` file:

```yml
steps:
  -
    name: Step 1
    command: echo hola
    agents:
      name: "$BUILDKITE_AGENT_META_DATA_NAME"
  -
    name: Step 2
    command: echo bonjour
    agents:
      name: "$BUILDKITE_AGENT_META_DATA_NAME"
  -
    name: Step 3
    command: echo konnichiwa
    agents:
      name: "$BUILDKITE_AGENT_META_DATA_NAME"
```

and the following executable script in `.buildkite/pipeline`:

```bash
#!/bin/bash

set -eu

# Makes sure all the steps run on this same agent
sed "s/\$BUILDKITE_AGENT_META_DATA_NAME/$BUILDKITE_AGENT_META_DATA_NAME/" pipeline.template.yml
```

it'd be neat if the `buildkite-agent pipeline upload` detected the presence of the executable file `.buildkite/pipeline` and executed it and read in it's STDIN, basically a default/defacto way of doing generated build pipelines the same as the default location for the yml file.

**Comment by toolmantim** [comment]
Also, stupid idea: a default `pipeline.template.yml` location that also-subs environment variables in

**Comment by keithpitt** [comment]
Another idea is to support a {{}} syntax, where ENV variables are read in:

```
  -
    name: Step 3
    command: echo konnichiwa
    agents:
      name: "{{$BUILDKITE_AGENT_META_DATA_NAME}}"
```

**Comment by lox** [comment]
Yeah, I'd be tempted to mirror the docker-compose syntax:
https://github.com/docker/compose/pull/1765

**Comment by keithpitt** [comment]
I was waiting for someone to come along with the perfect answer @lox :) It shall be done!

**Comment by keithpitt** [comment]
@lox I'm trying to find their current docs on the feature, to see if they handled the default value thing. Do you know if it handles default values?

**Comment by lox** [comment]
https://github.com/docker/compose/pull/1765#issuecomment-125617042

**Comment by lox** [comment]
tl;dr No, but they will support POSIX something something syntax: `${HOST_PORT-8080} or ${HOST_PORT:-8080}`

**Comment by keithpitt** [comment]
Yeah cool, this is ace. Will get it into the next release with that syntax, and I'll support the default gear swell.

**Comment by toolmantim** [comment]
🛰
